### PR TITLE
Make Problems immutable

### PR DESCRIPTION
**Pull request by hegyibalint**
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by donat** [comment]
@reinsch82 Please check the change set for `TODO` labels.

File: platforms/ide/problems-api/build.gradle.kts [hunk 1]
```diff
@@ -37,6 +37,7 @@ dependencies {
     implementation(libs.guava)
     implementation(libs.inject)
 
+    testImplementation(project(":logging"))
```
Comment (donat) [inline platforms/ide/problems-api/build.gradle.kts:40]: ❓ Is this needed? The contributed test seems to have no imports from the logging subproject.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java [hunk 2]
```diff
@@ -87,11 +87,19 @@ public interface Problem {
...
+    RuntimeException getException();
 
     /**
      * Additional Data about the problem.
      * In a map so it can be somewhat structured.
      */
     Map<String, String> getAdditionalData();
+
+    /**
+     * Returns a problem builder with fields initialized with values from this instance
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/Problem.java:99]:
```suggestion
     * Returns a problem builder with fields initialized with values from this instance.
```

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * Problem builder creating basic, non-reportable problems without enforcing any particular invocation order.
+ *
+ * @since 8.6
+ */
+@Incubating
+public interface UnboundBasicProblemBuilder extends UnboundProblemBuilder, BasicProblemBuilder {
+    UnboundBasicProblemBuilder documentedAt(DocLink doc);
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java:31]: 🤔 

We should add `inheritDoc` as documentation to all methods for all `Unbound*Builder` classes:

```
  /**
  * {@inheritDoc}
  */
```

File: platforms/ide/problems-api/src/test/groovy/org/gradle/api/problems/internal/DefaultReportableProblemTest.groovy [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        newProblem.label == problem.label
+        newProblem.additionalData == problem.additionalData
+        newProblem.details == problem.details
+        newProblem.exception == problem.exception
+        newProblem.locations == problem.locations
+        newProblem.severity == problem.severity
+        newProblem.equals(problem)
+
+        where:
+        severity         | additionalData
```
Comment (donat) [inline platforms/ide/problems-api/src/test/groovy/org/gradle/api/problems/internal/DefaultReportableProblemTest.groovy:51]: ❌ The variables are unused. 

- You may want to add another test where you modify a single field and check if was properly updated.
- Another tests coverage we should have:
  - reportableProbem.toBuilder().build().class == DefaultReportableProbelm
  - noReportableProbem.toBuilder().build().class == DefaultProbelm
  - reportableProbem.toBuilder().build().report() works, i.e. the mocked service is still available

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    UnboundBasicProblemBuilder undocumented();
+    UnboundBasicProblemBuilder fileLocation(String path, @Nullable Integer line, @Nullable Integer column, @Nullable Integer length);
+    UnboundBasicProblemBuilder pluginLocation(String pluginId);
+    UnboundBasicProblemBuilder stackLocation();
+    UnboundBasicProblemBuilder noLocation();
+    UnboundBasicProblemBuilder taskPathLocation(Path location);
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java:37]: `taskPathLocation` is not available on the the stepwise builder, so we are exposing something that's not even documented. I'd rather have some consistency. We have a few options:
- Adding taskPathLocation to the stepwise builder
- Removing taskPathLocation from the public api and fix the callsite with a cast to DefaultProblemBuilder.

**Review by donat — CHANGES_REQUESTED** [review]

File: platforms/ide/problems-api/build.gradle.kts (hunk 1 above)
Comment (reinsch82) [reply platforms/ide/problems-api/build.gradle.kts:40]: It's needed for `import org.gradle.internal.deprecation.Documentation`

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java (hunk 5 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/UnboundBasicProblemBuilder.java]: I removed the taskPathLocation from the public API

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle merge

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblem.java [hunk 6]
```diff
@@ -109,6 +111,10 @@ public Map<String, String> getAdditionalData() {
         return additionalMetadata;
     }
 
+    @Override
+    public UnboundBasicProblemBuilder toBuilder() {
+        return new DefaultBasicProblemBuilder(this);
+    }
```
Comment (alllex) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblem.java:117]: 💅 Looks like a newline is missing between methods

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultReportableProblem.java [hunk 7]
```diff
@@ -65,4 +66,25 @@ public void setProblemService(InternalProblems problemService) {
...
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        return this.problemService == ((DefaultReportableProblem) o).problemService && super.equals(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return problemService.hashCode() + super.hashCode();
```
Comment (alllex) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultReportableProblem.java:88]: 🤔 It's surprising to see a non-standard `hashCode` implementation. Would it have been possible to use something easily recognizable like `Objects.hashCode(...)` ?

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/InternalProblems.java [hunk 8]
```diff
@@ -30,14 +29,11 @@ public interface InternalProblems extends Problems {
...
      * The builder uses a stepwise builder pattern forcing the clients to define all mandatory fields in a specific order.
      * <p>
-     * Once all mandatory fields are set, the returned type will allow clients to call {@link BuildableProblemBuilder#build()} to create a new Problem instance.
-     * The {@link BuildableProblemBuilder#build()} method doesn't have any side effects, it just creates a new instance. Problems should be reported separately with {@link ReportableProblem#report()}.
+     * Once all mandatory fields are set, the returned type will allow clients to call {@link BasicProblemBuilder#build()} to create a new Problem instance.
+     * The {@link BasicProblemBuilder#build()} method doesn't have any side effects, it just creates a new instance. Problems should be reported separately with {@link ReportableProblem#report()}.
      *
      * @return a new problem builder
      */
     ProblemBuilderDefiningLabel createProblemBuilder();
```
Comment (alllex) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/InternalProblems.java:37]: 💅 A missing newline between methods?

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultReportableProblem.java (hunk 7 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultReportableProblem.java:88]: I overlooked those :(

I'll address it next time we touch it, which will be pretty soon :)
Thx for the feedback

**Pull request merged**